server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.server.jetty.jsplugin;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.deephaven.plugin.type.JsType;
+import io.deephaven.plugin.type.JsTypeRegistration;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletHolder;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link JsTypeRegistration} for Jetty.
+ *
+ * @see #servletHolder(String)
+ */
+public final class JsPlugins implements JsTypeRegistration {
+
+    /**
+     * Creates a new js plugins instance with a temporary directory.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final JsPlugins jsPlugins = new JsPlugins(Files.createTempDirectory(JsPlugins.class.getName()));
```

Comment:
We can't rely on py directories - they aren't guaranteed to be actual directories. This also allows us to eventually package stuff up in java jars if we need to; not generally accessible directly via the FS apis. (The resource abstraction that python presents uses a context manager / generator that allows temporary access to a pathlib.)

Reading the jetty documentation, they strongly recommend using DefaultServlet for static resources instead of Resource directly. I'll try and find that and link.

As it stands, we don't have dynamic support for our plugin infra. It's definitely a thing we can investigate in the future, but this isn't expected to work today. https://github.com/deephaven/deephaven-core/issues/2816

Reply:
It looks like even with `Resource.newResource(...)`, everything is *reference* based which we don't want when dealing with python resources.

I'm not sure if there is a pre-built "serve this content from an in-memory string". Of course, we could write that servlet logic ourselves.

Being able to take advantage of DefaultServlet is nice - my comment earlier about "they strongly recommend..." was a bit incorrect, but they do say this, https://www.eclipse.org/jetty/documentation/jetty-11/programming-guide/index.html#pg-server-http-handler-use-default-servlet:

> you may want to use a DefaultServlet instead of ResourceHandler

I'll continue digging to see if there is a good pre-existing solution for in-memory content.

As an alternative to temporary directory, we _can_ use the cache directory we setup as temporary space.

Reply:
Actually - `org.eclipse.jetty.util.resource.PathResource#PathResource(java.nio.file.Path)` seems like a solution if it actually works with non-filesystem "Path"s (ala something like https://github.com/google/jimfs). There is some hope that it "just works", given I see this in the constructor:

```java
this.belongsToDefaultFileSystem = this.path.getFileSystem() == FileSystems.getDefault();
```

Reply:
> Reading the jetty documentation, they strongly recommend using DefaultServlet for static resources instead of Resource directly. I'll try and find that and link.

Given that DefaultServlet wraps a Resource, and we already have a DefaultServlet, I would be surprised by this. We used to only set a directory for our resource base (not to be confused with base resource), but ... that causes jetty to unpack contents to temp, and that breaks when temp gets cleaned up.

Am I incorrect that installing a new wheel at runtime will not try to call register? should we check for calling register after init and then throw or ignore it?

Reply:
I've successfully used an in-memory Path (jimfs) so we don't need to deal w/ tmpfs.

Installing a wheel at runtime will not auto-register. That said, the raw plugin architecture (register(...)) _does_ support that functionality in theory, it just needs to be plumbed through. https://github.com/deephaven/deephaven-core/issues/2816